_posts/2020-07-12-ray-tune.md

```diff
@@ -0,0 +1,264 @@
+Let's look at each feature and see what Dask-ML and Ray have to offer:
+
+> 1\. **Consistency with Scikit-Learn API**: tune-sklearn is a drop-in
+> replacement for GridSearchCV and RandomizedSearchCV, so you only need to
+> change less than 5 lines in a standard Scikit-Learn script to use the API.
+
+Dask-ML is also consistent with the Scikit-learn API. Here's how to use
+Dask-ML's hyperparameter optimization:
+
+``` python
+from dask_ml.model_selection import HyperbandSearchCV
+from scipy.stats import loguniform
+from sklearn.datasets import make_classification
+from sklearn.linear_model import SGDClassifier
+
+X, y = make_classification()
+model = SGDClassifier()
+params = {"loss": ["hinge", "log", "modified_huber", "squared_hinge", "perceptron"],
+          "alpha": loguniform(1e-5, 1e-3)}
+
+search = HyperbandSearchCV(model, params)
+search.fit(X, y, classes=[0, 1])
+```
+
+This example is almost identical to [one of Ray's examples][sgd-eg] and [one of
+Scikit-learn's examples][skl-eg]. Clearly, Dask-ML offers a Scikit-learn
+compatible model selection API.
+
+[skl-eg]:https://scikit-learn.org/stable/auto_examples/model_selection/plot_randomized_search.html#sphx-glr-auto-examples-model-selection-plot-randomized-search-py
+[sgd-eg]:https://github.com/ray-project/tune-sklearn/blob/31f228e21ef632a89a74947252d8ad5323cbd043/examples/sgd.py
+
+> 2\. **Modern hyperparameter tuning techniques:** tune-sklearn is the only
+> Scikit-Learn interface that allows you to easily leverage Bayesian
+> Optimization, HyperBand, and other optimization techniques by simply toggling
+> a few parameters.
+
+Dask-ML also offers modern hyperparameter tuning techniques. Hyperband is the
+most prominent example; after the introduction in 2016, [the paper by Li et.
+al][hyperband-paper] has been cited [over 470 times][470] and has been
+implemented in many different libraries including [Dask-ML][hscv],
+[keras-tune], [Optuna], AutoML[^automl] and [Microsoft's NNI][nni].
+
+Hyperband is so popular because it's so effective. In the introduction paper by
+Li et.  al, they showed a rather drastic improvement when comparing with all
+the relevant implementations,[^hyperband-figs] and this drastic improvement has
+remained in follow-up works.[^follow-up]
+
+
+[^follow-up]:See Figure 1 of [the BOHB paper][BOHB paper] and [a paper][blippar] from an augmented reality company.
+
+[^hyperband-figs]:See Figures 4, 7 and 8 in "[Hyperband: A Novel Bandit-Based Approach toHyperparameter Optimization][hyperband-paper]."
+
+[blippar]:https://arxiv.org/pdf/1801.01596.pdf
+[BOHB paper]:http://proceedings.mlr.press/v80/falkner18a/falkner18a.pdf
+[nni]:https://nni.readthedocs.io/en/latest/Tuner/HyperbandAdvisor.html
+
+Even Bayesian hyperparameter optimization have been adapted to the Hyperband
+implementation in AutoML's [BOHB]. Dask-ML could use more Bayesian sampling
+approaches; however, this is difficult given the distributed nature of Dask and
```

Comment:
Is there anything specific to Dask's architecture that makes Bayesian methods hard? Or is a generally hard thing for distributed systems to do. If it's general, then replace "Dask" with something like "any distributed system".

Reply:
Thanks!